### ISSUE TITLE
recurly.configure does not reinject iframes.

### ISSUE THREAD

**Issue by lukebrdn** [recurly/recurly-js#253]
I'm building a form to update billing information. The user can reveal or hide this form by clicking an edit button. The first time recurly.configure is called, the iframes inject correctly. However, if I destroy the form view and then the user reopens the form, which calls recurly.configure again, the iframes will not be injected.

I've solved the issue by setting recurly.readyState = 0, whenever the user hides the form. This feels a little hacky, so it'd be nice if there was some type of reset function, or if recurly.configure could check if the iframes were present rather than rely on the readyState.

**Comment by chrissrogers** [comment]
#243 is meant to address this. It's marked WIP because there is a readyState race condition present during first initialization. I imagine a complete solution will involve a better state tracking than `readyState` provides..

You're right that manipulating readyState is a bit of a hack, since the orphaned message bus will continue to listen to and send messages, though they won't have attentive receivers.

**Comment by bennadel** [comment]
I've also gone the `readyState = 0` approach, which works in Chrome, Firefox, and Safari. However when I try to use this approach in **IE**, I get a JavaScript error when re-showing the form:

```
Invalid operand to 'instanceof': Function expected
File: hosted-field.js, Line: 4, Column: 322
```

However, when I click into the field, it updates the CSS classes appropriately (in IE); but, it won't let me type into the text fields. I have to refresh the page to make the form interactive again.

**Comment by bennadel** [comment]
Sorry, just a quick follow-up. I get the above JavaScript error in IE _whether or not_ the `readyState` code is in there. It looks like IE simply can't re-render the hosted fields properly.

**Comment by marbemac** [comment]
Any progress on this / does this hack still work? Can't seem to get it to work for some reason.

**Comment by bennadel** [comment]
For what it's worth, we started looking at using an IFrame to render/destroy the Recurly.js rendering as we could not get the re-injection to work in IE.

**Comment by chrissrogers** [comment]
Progress on this is reflected in #267 -- the scope is a bit bigger than the original method, but much cleaner altogether. There is a currently a bug in initial and reconfiguration with multiple instances. Once that's worked out, this will be all remedied.

**Comment by queso** [comment]
@chrissrogers: what is the ETA on that fix landing? We have a pending production launch and this bug hits us when we use Recurly with React.

**Comment by marbemac** [comment]
This also hits us in production right now with React. We have various workarounds and hacks to try to mitigate the issue, but it still pops up in certain situations.

**Comment by queso** [comment]
@marbemac how are you working around it atm? We just starting to move all the recurly stuff into it's own iFrame and then I was going to attach some event listeners to that iFrame for it.

**Comment by marbemac** [comment]
Are you building w Meteor? We tried the iframe route and ultimately decided loading the entire meteor app again inside of it was not quick enough. We show the billing capture form in a modal, and delay loading recurly js until that modal is loaded (react-script-loader). We also don't destroy this particular billing modal component once it's been mounted, but rather just hide it on the page w css, high up in the dom. The combination of these two measures means that once rendered, the component won't be unmounted, and thus won't need to go through the whole recurly config process. As a last resort, in the modal, we have a chat button and note to click it if anything is not working.

**Comment by queso** [comment]
Yeah, we have a high degree of mobile and want to avoid a modal. Trying to a static iframe for now. Honestly, I may just end up with a forced browser refresh on the page or something.

**Comment by marbemac** [comment]
Right well the concept would be similar. The component that you use to capture billing - the one that loads and configures recurly, needs to not be unmounted. Instead, you can try hiding it somewhere high in the tree. It could be a full screen component that takes up the entire mobile viewport, thus looking like a regular page. The key is to never unmount it once it's been mounted.

**Comment by chrissrogers** [comment]
I don't have a concrete ETA for you guys unfortunately. If I get a better sense I will certainly pop an update in here.

**Comment by bennadel** [comment]
Yesterday, we launched part of our IFrame-based solution. We basically pass in all the existing form-data in the IFrame URL, then parse it, render the form (in the IFrame), and .configure() the Recurly.js fields. 

In order to get all this to work, we had to create a "proxy" around the IFrame that knows how to render the IFrame with the parent window when the IFrame loads; and then how to de-register everything when the IFrame is destroyed.

It was incredibly __not-trivial__ to get it working, unfortunately.

**Comment by bennadel** [comment]
@queso we also explored the "just refresh the page" option. The problem we were running into is that this would be required whether or not the form was actually submitted. Meaning, imagine the user opens the form, then cancels out. This created -- at least for us -- all the same problems. As such, we would have had to refresh the page even when the user didn't truly engage with the form; which seems like a bad UX.

**Comment by volodymyrrudyi** [comment]
same issue here, it makes Recurly.js almost unusable in SPAs without implementing lots of workarounds

**Comment by geneparcellano** [comment]
We had this same issue in our Angular app, using ui-router. We were able to solve it using [http://christopherthielen.github.io/ui-router-extras/#/sticky](http://christopherthielen.github.io/ui-router-extras/#/sticky).

**Comment by venuswei** [comment]
Faced the same problem with Ember apps as well.. The form loads alright when it is first initialised on that page, but once we transition to another page and comes back, the form is gone. Any solutions yet?

**Comment by marbemac** [comment]
This is a bug that affects pretty much every recurly customer that uses a modern client-side js lib (ember, angular, react, blaze, backbone, etc). @chrissrogers any updates? No ill-will intended (we love recurly and recurly-js!) but frankly I'm surprised that this issue hasn't received more attention.

**Comment by developermhayden** [comment]
I had the same issue but a colleague pointed out that I was using an ng-switch rather than an ng-show when showing my Recurly form in our SPA.
ng-switch doesn't put the Recurly form in the dom.
ng-show puts the Recurly form in the DOM but hides it's visibility.

Once I switched to using ng-show rather than ng-switch, Recurly form worked correctly navigating away and back to the form on our SPA.

**Comment by chrissrogers** [comment]
@marbemac sorry thought I had replied to you last week! I'm prioritizing this right now. It involves a rearchitecting of the instantiation process, so more updates incoming!

**Comment by chrissrogers** [comment]
This will be available in [v4.3.0](https://github.com/recurly/recurly-js/releases/tag/v4.3.0), which is slated for release next week. You'll be able to call `recurly.configure` and specify new selectors for your fields to target at any time during the page lifecycle.

**Comment by chrissrogers** [comment]
Just a note that the previously-discussed workaround of setting `recurly.resdyState = 0` will cause the hosted fields to fail to load in `v4.3.0+`
